File: src/types/analytics.ts
// ─────────────────────────────────────────────────────────────────────────────
// Phase 6.5 — Analytics Platform (Metrics / Snapshots / Reports)
// ─────────────────────────────────────────────────────────────────────────────

export type MetricScope =
  | "volunteer"
  | "disaster"
  | "public_resource"
  | "translation"
  | "organization"
  | "platform";

export interface AnalyticsMetric {
  id: string;
  metricName: string;
  metricScope: MetricScope | string;
  entityType: string | null;
  entityId: string | null;
  metricValue: number;
  metricUnit: string | null;
  recordedAt: string;
  metadata: Record<string, unknown>;
  createdAt: string;
  updatedAt: string;
}

export interface AnalyticsMetricInput {
  metricName: string;
  metricScope: MetricScope | string;
  metricValue: number;
  recordedAt: string;
  entityType?: string | null;
  entityId?: string | null;
  metricUnit?: string | null;
  metadata?: Record<string, unknown>;
}

export interface MetricListQuery {
  q?: string;
  metricScope?: string;
  metricName?: string;
  entityType?: string;
  entityId?: string;
  recordedFrom?: string;
  recordedTo?: string;
  page?: number;
  pageSize?: number;
}

export type SnapshotType = "daily" | "weekly" | "monthly" | "custom";

export interface AnalyticsSnapshot {
  id: string;
  snapshotType: SnapshotType | string;
  organizationId: string | null;
  periodStart: string;
  periodEnd: string;
  generatedAt: string;
  metricsJson: Record<string, unknown>;
  metadata: Record<string, unknown>;
  createdAt: string;
  updatedAt: string;
}

export interface AnalyticsSnapshotInput {
  snapshotType: SnapshotType | string;
  periodStart: string;
  periodEnd: string;
  metricsJson?: Record<string, unknown>;
  organizationId?: string | null;
  metadata?: Record<string, unknown>;
}

export type ReportStatus = "pending" | "generating" | "completed" | "failed" | "expired";

export interface AnalyticsReport {
  id: string;
  reportName: string;
  reportType: string;
  requestedByUserId: string | null;
  organizationId: string | null;
  status: ReportStatus | string;
  filePath: string | null;
  generatedAt: string | null;
  expiresAt: string | null;
  metadata: Record<string, unknown>;
  createdAt: string;
  updatedAt: string;
}

export interface AnalyticsReportInput {
  reportName: string;
  reportType: string;
  requestedByUserId?: string | null;
  organizationId?: string | null;
  metadata?: Record<string, unknown>;
}

export interface SnapshotListQuery {
  snapshotType?: string;
  organizationId?: string;
  page?: number;
  pageSize?: number;
}

export interface ReportListQuery {
  status?: string;
  reportType?: string;
  organizationId?: string;
  requestedByUserId?: string;
  page?: number;
  pageSize?: number;
}

export interface MetricAggregate {
  count: number;
  sum: number;
  avg: number;
  min: number;
  max: number;
}

export interface PlatformOverviewKpis {
  totalVolunteers: number;
  activeDisasters: number;
  publicResources: number;
  publishedTranslations: number;
  organizations: number;
  reportsGenerated: number;
}

export interface KpiPoint {
  label: string;
  value: number;
  delta?: number;
  helper?: string;
}

export interface TimeSeriesPoint {
  date: string;
  [metric: string]: string | number;
}

export interface DistributionSlice {
  label: string;
  value: number;
  color?: string;
}

export interface ExecutiveOverview {
  kpis: KpiPoint[];
  reach: TimeSeriesPoint[];
  audienceGrowth: TimeSeriesPoint[];
  languages: DistributionSlice[];
  organizations: DistributionSlice[];
  engagement: TimeSeriesPoint[];
  delivery: TimeSeriesPoint[];
  activity: TimeSeriesPoint[];
  approvals: KpiPoint[];
}

export type ReportKind =
  | "campaign"
  | "audience"
  | "organization"
  | "delivery"
  | "template"
  | "translation"
  | "ai"
  | "audit"
  | "security"
  | "activity";

export interface SavedReport {
  id: string;
  name: string;
  kind: ReportKind;
  description: string;
  createdBy: string;
  createdAt: string;
  lastRunAt?: string;
  filters: Record<string, string | number | string[] | undefined>;
  scheduled?: boolean;
}

export interface ReportRow {
  id: string;
  [column: string]: string | number;
}

export interface ReportResult {
  columns: { key: string; label: string; kind: "text" | "number" | "date" }[];
  rows: ReportRow[];
  total: number;
}

/* ================================================================== */
/* Phase 6 — Platform Analytics & Reporting                            */
/* ------------------------------------------------------------------ */
/* Single aggregate model for the unified reporting center. Every    */
/* module surface (campaigns, qr, volunteers, disasters, multilingual */
/* public) is a namespaced block on the same response so the UI can  */
/* filter it in a single query. Backend contract lives in            */
/* docs/BACKEND-ANALYTICS.md.                                        */
/* ================================================================== */

export type AnalyticsScope = "platform" | "organization" | "personal";

export interface AnalyticsFilters {
  from?: string; // ISO date
  to?: string;   // ISO date
  campaignIds?: string[];
  disasterIds?: string[];
  organizationIds?: string[];
  languages?: string[];
  volunteerIds?: string[];
}

export interface CampaignAnalytics {
  totals: {
    total: number;
    published: number;
    draft: number;
    archived: number;
    reach: number;
    completion: number;    // 0..1 — average completion
    engagement: number;    // 0..1 — average engagement rate
    downloads: number;
  };
  top: {
    id: string;
    name: string;
    reach: number;
    engagement: number;
  }[];
  timeline: TimeSeriesPoint[]; // { date, published, archived }
}

export interface QrAnalyticsAggregate {
  totals: {
    scans: number;
    unique: number;
    repeat: number;
  };
  byCountry: DistributionSlice[];
  byLanguage: DistributionSlice[];
  byDevice: DistributionSlice[];
  trend: TimeSeriesPoint[]; // { date, scans, unique }
}

export interface VolunteerAnalytics {
  totals: {
    registered: number;
    available: number;
    assigned: number;
    completedTasks: number;
    averageCompletionHours: number;
  };
  activity: TimeSeriesPoint[]; // { date, assigned, completed }
  topContributors: {
    id: string;
    name: string;
    completedTasks: number;
    hoursContributed: number;
  }[];
}

export interface DisasterAnalytics {
  totals: {
    active: number;
    resolved: number;
    volunteersAssigned: number;
    emergencyCampaigns: number;
    publicAlertReach: number;
    averageResponseHours: number;
  };
  bySeverity: DistributionSlice[];
  timeline: TimeSeriesPoint[]; // { date, opened, resolved }
}

export interface MultilingualAnalytics {
  totals: {
    languagesUsed: number;
    translationsGenerated: number;
    translationsPublished: number;
    audioGenerated: number;
    audioPlays: number;
    coverage: number; // 0..1 — percent of entities with >=1 non-source language published
  };
  byLanguage: DistributionSlice[]; // count of published translations per language
  byStatus: DistributionSlice[];   // draft/generated/edited/published
  mostUsedLanguage?: string;
}

/** Personal contribution snapshot for the Volunteer analytics dashboard. */
export interface PersonalAnalytics {
  totals: {
    assignedTasks: number;
    completedTasks: number;
    hoursContributed: number;
    campaignsSupported: number;
    disastersSupported: number;
    languagesContributed: number;
  };
  activity: TimeSeriesPoint[]; // { date, completed }
  recent: {
    id: string;
    title: string;
    kind: "task" | "translation" | "audio";
    completedAt: string;
  }[];
}

/** Unified platform analytics response — one call per dashboard render. */
export interface PlatformAnalytics {
  scope: AnalyticsScope;
  range: { from: string; to: string };
  campaigns: CampaignAnalytics;
  qr: QrAnalyticsAggregate;
  volunteers: VolunteerAnalytics;
  disasters: DisasterAnalytics;
  multilingual: MultilingualAnalytics;
  personal?: PersonalAnalytics; // present when scope="personal"
}

export type ExportFormat = "csv" | "xlsx" | "pdf" | "json";

export type ReportDataset =
  | "campaigns"
  | "qr"
  | "volunteers"
  | "disasters"
  | "multilingual"
  | "public_engagement"
  | "platform";

export interface ExportRequest {
  dataset: ReportDataset;
  format: ExportFormat;
  filters: AnalyticsFilters;
}

/** Signed URL response — the frontend never streams binary data itself. */
export interface ExportResult {
  url: string;
  format: ExportFormat;
  dataset: ReportDataset;
  expiresAt: string;
  sizeBytes?: number;
}

File: backend/app/schemas/analytics.py
"""Pydantic schemas for the Analytics & Reporting platform (Phase 6.1).

Legacy KPI/report DTOs used by the existing analytics service are
preserved. New DTOs power the AnalyticsMetric / AnalyticsSnapshot /
AnalyticsReport tables and mirror :mod:`app.constants.analytics`.
"""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from app.schemas.common import IdentifiedDto


# -- Legacy analytics DTOs (unchanged) ---------------------------------------


class AnalyticsKpiDto(BaseModel):
    key: str
    label: str
    value: float
    delta: float | None = None
    unit: str | None = None


class ReportDto(IdentifiedDto):
    name: str
    kind: str
    scheduled: bool = False


# -- Enum literals (mirror app.constants.analytics) --------------------------

MetricScope = Literal[
    "volunteer",
    "disaster",
    "public_resource",
    "translation",
    "organization",
    "platform",
]
ReportStatus = Literal["pending", "generating", "completed", "failed"]
SnapshotType = Literal["daily", "weekly", "monthly", "custom"]


# --------------------------------------------------------------------------- #
# AnalyticsMetric
# --------------------------------------------------------------------------- #


class AnalyticsMetricBase(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    metricName: str = Field(alias="metric_name", min_length=1, max_length=120)
    metricScope: MetricScope = Field(alias="metric_scope")
    entityType: str | None = Field(alias="entity_type", default=None, max_length=40)
    entityId: uuid.UUID | None = Field(alias="entity_id", default=None)
    metricValue: float = Field(alias="metric_value", default=0.0)
    metricUnit: str | None = Field(alias="metric_unit", default=None, max_length=40)


class AnalyticsMetricCreate(AnalyticsMetricBase):
    recordedAt: datetime = Field(alias="recorded_at")
    metadata: dict[str, Any] = Field(default_factory=dict)


class AnalyticsMetricUpdate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    metricValue: float | None = Field(alias="metric_value", default=None)
    metricUnit: str | None = Field(alias="metric_unit", default=None, max_length=40)
    metadata: dict[str, Any] | None = None


class AnalyticsMetricDto(IdentifiedDto):
    metricName: str = Field(alias="metric_name")
    metricScope: MetricScope = Field(alias="metric_scope")
    entityType: str | None = Field(alias="entity_type", default=None)
    entityId: uuid.UUID | None = Field(alias="entity_id", default=None)
    metricValue: float = Field(alias="metric_value")
    metricUnit: str | None = Field(alias="metric_unit", default=None)
    recordedAt: datetime = Field(alias="recorded_at")
    metadata: dict[str, Any] = Field(default_factory=dict)


class AnalyticsMetricListQuery(BaseModel):
    metricScope: MetricScope | None = None
    metricName: str | None = None
    entityType: str | None = None
    entityId: uuid.UUID | None = None
    recordedFrom: datetime | None = None
    recordedTo: datetime | None = None
    page: int = 1
    pageSize: int = 50


# --------------------------------------------------------------------------- #
# AnalyticsSnapshot
# --------------------------------------------------------------------------- #


class AnalyticsSnapshotBase(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    snapshotType: SnapshotType = Field(alias="snapshot_type")
    organizationId: uuid.UUID | None = Field(alias="organization_id", default=None)
    periodStart: datetime = Field(alias="period_start")
    periodEnd: datetime = Field(alias="period_end")

    @model_validator(mode="after")
    def _validate_period(self):
        if self.periodEnd < self.periodStart:
            raise ValueError("period_end must be >= period_start")
        return self


class AnalyticsSnapshotCreate(AnalyticsSnapshotBase):
    generatedAt: datetime = Field(alias="generated_at")
    metricsJson: dict[str, Any] = Field(alias="metrics_json", default_factory=dict)
    metadata: dict[str, Any] = Field(default_factory=dict)


class AnalyticsSnapshotUpdate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    metricsJson: dict[str, Any] | None = Field(alias="metrics_json", default=None)
    metadata: dict[str, Any] | None = None


class AnalyticsSnapshotDto(IdentifiedDto):
    snapshotType: SnapshotType = Field(alias="snapshot_type")
    organizationId: uuid.UUID | None = Field(alias="organization_id", default=None)
    periodStart: datetime = Field(alias="period_start")
    periodEnd: datetime = Field(alias="period_end")
    generatedAt: datetime = Field(alias="generated_at")
    metricsJson: dict[str, Any] = Field(alias="metrics_json", default_factory=dict)
    metadata: dict[str, Any] = Field(default_factory=dict)


# --------------------------------------------------------------------------- #
# AnalyticsReport
# --------------------------------------------------------------------------- #


class AnalyticsReportBase(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    reportName: str = Field(alias="report_name", min_length=1, max_length=200)
    reportType: str = Field(alias="report_type", min_length=1, max_length=40)
    organizationId: uuid.UUID | None = Field(alias="organization_id", default=None)


class AnalyticsReportCreate(AnalyticsReportBase):
    requestedByUserId: uuid.UUID | None = Field(
        alias="requested_by_user_id", default=None
    )
    status: ReportStatus = "pending"
    metadata: dict[str, Any] = Field(default_factory=dict)


class AnalyticsReportUpdate(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    status: ReportStatus | None = None
    filePath: str | None = Field(alias="file_path", default=None, max_length=500)
    generatedAt: datetime | None = Field(alias="generated_at", default=None)
    expiresAt: datetime | None = Field(alias="expires_at", default=None)
    metadata: dict[str, Any] | None = None


class AnalyticsReportDto(IdentifiedDto):
    reportName: str = Field(alias="report_name")
    reportType: str = Field(alias="report_type")
    requestedByUserId: uuid.UUID | None = Field(
        alias="requested_by_user_id", default=None
    )
    organizationId: uuid.UUID | None = Field(alias="organization_id", default=None)
    status: ReportStatus
    filePath: str | None = Field(alias="file_path", default=None)
    generatedAt: datetime | None = Field(alias="generated_at", default=None)
    expiresAt: datetime | None = Field(alias="expires_at", default=None)
    metadata: dict[str, Any] = Field(default_factory=dict)


class AnalyticsReportListQuery(BaseModel):
    status: ReportStatus | None = None
    reportType: str | None = None
    organizationId: uuid.UUID | None = None
    requestedByUserId: uuid.UUID | None = None
    generatedFrom: datetime | None = None
    generatedTo: datetime | None = None
    page: int = 1
    pageSize: int = 50


class AnalyticsSearchQuery(BaseModel):
    """Free-text search across analytics reports / snapshots."""

    q: str | None = None
    metricScope: MetricScope | None = None
    snapshotType: SnapshotType | None = None
    reportStatus: ReportStatus | None = None
    page: int = 1
    pageSize: int = 50
